src/execution_plan/ops/op_index_scan.c

```diff
@@ -52,12 +51,19 @@ Record IndexScanConsume(OpBase *opBase) {
 }
 
 OpResult IndexScanReset(OpBase *ctx) {
-	IndexScan *indexScan = (IndexScan *)ctx;
-	IndexIter_Reset(indexScan->iter);
+	IndexScan *op = (IndexScan *)ctx;
+	RediSearch_ResultsIteratorReset(op->iter);
 	return OP_OK;
 }
 
-void IndexScanFree(OpBase *op) {
-	IndexScan *indexScan = (IndexScan *)op;
-	IndexIter_Free(indexScan->iter);
+void IndexScanFree(OpBase *ctx) {
+	IndexScan *op = (IndexScan *)ctx;
+	/* As long as this Index iterator is alive the index is
+	 * read locked, if this index scan operation is part of
+	 * a query which will modified this index we'll be stuck in
+	 * a dead lock, as we're unable to acquire index write lock. */
```

Comment:
We can still get deadlocked in some scenarios with multiple streams:
```
$ ge social "MATCH (p:person) SET p.name = 'f' WITH p MATCH (q:person) WHERE q.name > '' RETURN p, q"
1) "Results"
2) "    Project"
3) "        Cartesian Product"
4) "            Project"
5) "                Update"
6) "                    Node By Label Scan | (p:person)"
7) "            Index Scan | (q:person)"
```
This will get into a deadlock with the backtrace:
```
#0  0x00007ffff74189e4 in futex_abstimed_wait (private=<optimized out>, abstime=0x0, expected=2, futex_word=0x7ffff4cd2de8 <RWLock+8>)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:172
#1  __pthread_rwlock_wrlock_full (abstime=0x0, rwlock=0x7ffff4cd2de0 <RWLock>) at pthread_rwlock_common.c:803
#2  __GI___pthread_rwlock_wrlock (rwlock=0x7ffff4cd2de0 <RWLock>) at pthread_rwlock_wrlock.c:27
#3  0x00007ffff47a6a70 in RediSearch_DeleteDocument (sp=0x7ffff6c9b400, docKey=0x7ffff2e44360, len=8)
    at /home/tower-linux/dev/redis/graph/deps/RediSearch/src/redisearch_api.c:142
#4  0x00007ffff4723295 in Index_RemoveNode (idx=idx@entry=0x7ffff6c0f930, n=n@entry=0x7fffef22b010)
    at /home/tower-linux/dev/redis/graph/src/index/index.c:170
#5  0x00007ffff472d8cb in Schema_AddNodeToIndices (s=s@entry=0x7ffff6c0e1c0, n=n@entry=0x7fffef22b010, update=update@entry=true)
    at /home/tower-linux/dev/redis/graph/src/schema/schema.c:115
#6  0x00007ffff470e6d4 in _UpdateIndex (gc=<optimized out>, old_value=<optimized out>, new_value=0x7fffef22b078, s=0x7ffff6c0e1c0, ctx=0x7fffef22b000)
    at /home/tower-linux/dev/redis/graph/src/execution_plan/ops/op_update.c:44
#7  _UpdateNode (op=0x7fffd8202630, ctx=0x7fffef22b000) at /home/tower-linux/dev/redis/graph/src/execution_plan/ops/op_update.c:73
#8  _CommitUpdates (op=0x7fffd8202630) at /home/tower-linux/dev/redis/graph/src/execution_plan/ops/op_update.c:107
#9  OpUpdateConsume (opBase=0x7fffd8202630) at /home/tower-linux/dev/redis/graph/src/execution_plan/ops/op_update.c:209
```

Reply:
Interesting, need to investigate

Reply:
the `Update` operation should be relocated right below `Cartesian Product`
see Neo's execution plan as a reference.

Reply:
Or introduce the eager operation (which might not solve our issue)
https://neo4j.com/docs/cypher-manual/current/execution-plans/operators/#query-plan-eager